src/Parsing/Impl/FStringParser.cs

```diff
@@ -0,0 +1,147 @@
+﻿using System;
+using System.IO;
+using System.Text;
+using Microsoft.Python.Core;
+using Microsoft.Python.Core.Diagnostics;
+using Microsoft.Python.Core.Text;
+using Microsoft.Python.Parsing.Ast;
+
+namespace Microsoft.Python.Parsing {
+    public class FStringParser {
+        private readonly string _fString;
+        private readonly StringBuilder _buffer = new StringBuilder();
+        private readonly FStringBuilder _builder;
+        private readonly ErrorSink _errors;
+        private int _errorCode = 0;
+        private int _position = 0;
+
+        public FStringParser(FStringBuilder builder, string fString, ErrorSink errors) {
+            _fString = fString;
+            _builder = builder;
+            _errors = errors;
+        }
+
+        public void Parse() {
+            while (!EndOfFString()) {
+                if (IsDoubleBrace()) {
+                    NextChar();
+                    _buffer.Append(NextChar());
+                } else if (CurrentChar() == '{') {
+                    AddBufferedSubstring();
+                    ParseInnerExpression();
+                } else if (CurrentChar() == '}') {
+                    throw new Exception("closing '}' without opening");
+                } else {
+                    _buffer.Append(NextChar());
+                }
+            }
+            AddBufferedSubstring();
+        }
+
+        private bool IsDoubleBrace() {
+            StringSpan doubleOpen = new StringSpan("{{", 0, 2);
+            StringSpan doubleClose = new StringSpan("}}", 0, 2);
+            return IsNext(doubleOpen) || IsNext(doubleClose);
+        }
+
+        private bool IsNext(StringSpan span) {
+            return _fString.Slice(_position, span.Length).Equals(span);
+        }
+
+        private void ParseInnerExpression() {
+            Check.InvalidOperation(_buffer.Length == 0, "Current buffer is not empty");
+
+            Read('{');
+            AppendAllSubExpression();
+            Read('}');
+
+            var subExprStr = _buffer.ToString();
+            if (!subExprStr.IsNullOrEmpty()) {
+                _builder.AppendExpression(CreateExpression(_buffer.ToString()));
+            }
+            _buffer.Clear();
+        }
+
+        private Expression CreateExpression(string subExprStr) {
+            Check.ArgumentNotNullOrEmpty(nameof(subExprStr), subExprStr);
+
+            var parser = Parser.CreateParser(new StringReader(subExprStr), PythonLanguageVersion.V36);
```

Comment:
Hardcoded language version

Reply:
Speaking of, if we discover f-string in Python 2 we should report that the syntax requires Python 3 since parser does this for other new constructs